# Dealing with Messy Data
>  This chapter introduces you to the reality of messy and incomplete data. You will learn how to find where your data has missing values and explore multiple approaches on how to deal with them. You will also use string manipulation techniques to deal with unwanted characters in your dataset.

- toc: true 
- badges: true
- comments: true
- author: Lucas Nunes
- categories: [Python, Datacamp, Machine Learning]
- image: images/datacamp/1_supervised_learning_with_scikit_learn/2_regression.png

> Note: This is a summary of the course's chapter 2 exercises "Feature Engineering for Machine Learning in Python" at datacamp. <br>[Github repo](https://github.com/lnunesAI/Datacamp/) / [Course link](https://www.datacamp.com/tracks/machine-learning-scientist-with-python)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Why do missing values exist?

### How sparse is my data?

<div class=""><p>Most data sets contain missing values, often represented as NaN (Not a Number). If you are working with Pandas you can easily check how many missing values exist in each column.</p>
<p>Let's find out how many of the developers taking the survey chose to enter their age (found in the <code>Age</code> column of <code>so_survey_df</code>) and their gender (<code>Gender</code> column of <code>so_survey_df</code>).</p></div>

In [25]:
so_survey_df = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/10-feature-engineering-for-machine-learning-in-python/datasets/so_survey_df.csv')

Instructions 1/2
<ul>
<li>Subset the DataFrame to only include the <code>'Age'</code> and <code>'Gender'</code> columns. </li>
<li>Print the number of non-missing values in both columns.</li>
</ul>

In [26]:
# Subset the DataFrame
sub_df = so_survey_df[['Age', 'Gender']]

# Print the number of non-missing values
print(sub_df.notnull().sum())

Age       999
Gender    693
dtype: int64


Instructions 2/2
<p>Based on the results, how many non-missing entries are there in the <code>Gender</code> column?</p>

<pre>
Possible Answers

999

<b>693</b>

</pre>

**there are 693 non-missing entries in the Gender column.**

### Finding the missing values

<p>While having a summary of how much of your data is missing can be useful, often you will need to find the exact locations of these missing values.
Using the same subset of the StackOverflow data from the last exercise (<code>sub_df</code>), you will show how a value can be flagged as missing.</p>

Instructions 1/3
<p>Print the first 10 entries of the DataFrame.</p>

In [27]:
# Print the top 10 entries of the DataFrame
print(sub_df.head(10))

   Age  Gender
0   21    Male
1   38    Male
2   45     NaN
3   46    Male
4   39    Male
5   39    Male
6   34    Male
7   24  Female
8   23    Male
9   36     NaN


Instructions 2/3
<p>Print the locations of the missing values in the first 10 rows.</p>

In [28]:
# Print the locations of the missing values
sub_df.head(10).isna()

Age  Gender
0  False   False
1  False   False
2  False    True
3  False   False
4  False   False
5  False   False
6  False   False
7  False   False
8  False   False
9  False    True

Instructions 3/3
<p>Print the locations of the non-missing values in the first 10 rows.</p>

In [29]:
# Print the locations of the non-missing values
sub_df.head(10).notna()

Age  Gender
0  True    True
1  True    True
2  True   False
3  True    True
4  True    True
5  True    True
6  True    True
7  True    True
8  True    True
9  True   False

### Listwise deletion

<div class=""><p>The simplest way to deal with missing values in your dataset when they are occurring entirely at random is to remove those rows, also called 'listwise deletion'.</p>
<p>Depending on the use case, you will sometimes want to remove all missing values in your data while other times you may want to only remove a particular column if too many values are missing in that column.</p></div>

Instructions 1/4
<p>Print the number of rows and columns in <code>so_survey_df</code>.</p>

In [30]:
# Print the number of rows and columns
print(so_survey_df.shape)

(999, 11)


Instructions 2/4
<p>Drop all rows with missing values in <code>so_survey_df</code>.</p>

In [31]:
# Create a new DataFrame dropping all incomplete rows
no_missing_values_rows = so_survey_df.dropna()

# Print the shape of the new DataFrame
print(no_missing_values_rows.shape)

(264, 11)


Instructions 3/4
<p>Drop all columns with missing values in <code>so_survey_df</code>.</p>

In [32]:
# Create a new DataFrame dropping all columns with incomplete rows
no_missing_values_cols = so_survey_df.dropna(axis=1)

# Print the shape of the new DataFrame
print(no_missing_values_cols.shape)

(999, 7)


Instructions 4/4
<p>Drop all rows in <code>so_survey_df</code> where <code>'Gender'</code> is missing.</p>

In [33]:
# Drop all rows where Gender is missing
no_gender = so_survey_df.dropna(subset=['Gender'])

# Print the shape of the new DataFrame
print(no_gender.shape)

(693, 11)


**as you can see dropping all rows that contain any missing values may greatly reduce the size of your dataset. So you need to think carefully and consider several trade-offs when deleting missing values.**

### Replacing missing values with constants

<div class=""><p>While removing missing data entirely maybe a correct approach in many situations, this may result in a lot of information being omitted from your models.</p>
<p>You may find categorical columns where the missing value is a valid piece of information in itself, such as someone refusing to answer a question in a survey.
In these cases, you can fill all missing values with a new category entirely, for example 'No response given'.</p></div>

Instructions 1/2
<p>Print the count of occurrences of each category in <code>so_survey_df</code>'s <code>Gender</code> column.</p>

In [34]:
# Print the count of occurrences
print(so_survey_df['Gender'].value_counts())

Male                                                                         632
Female                                                                        53
Transgender                                                                    2
Female;Male                                                                    2
Female;Transgender                                                             1
Non-binary. genderqueer. or gender non-conforming                              1
Male;Non-binary. genderqueer. or gender non-conforming                         1
Female;Male;Transgender;Non-binary. genderqueer. or gender non-conforming      1
Name: Gender, dtype: int64


Instructions 2/2
<p>Replace all missing values in the <code>Gender</code> column with the string <code>'Not Given'</code>. Make changes to the original DataFrame.</p>

In [35]:
# Replace missing values
so_survey_df['Gender'].fillna('Not Given', inplace=True)

# Print the count of each value
print(so_survey_df['Gender'].value_counts())

Male                                                                         632
Not Given                                                                    306
Female                                                                        53
Transgender                                                                    2
Female;Male                                                                    2
Female;Transgender                                                             1
Non-binary. genderqueer. or gender non-conforming                              1
Male;Non-binary. genderqueer. or gender non-conforming                         1
Female;Male;Transgender;Non-binary. genderqueer. or gender non-conforming      1
Name: Gender, dtype: int64


**By filling in these missing values you can use the columns in your analyses.**

## Dealing with missing values (II)

### Filling continuous missing values

<p>In the last lesson, you dealt with different methods of removing data missing values and filling in missing values with a fixed string. These approaches are valid in many cases, particularly when dealing with categorical columns but have limited use when working with continuous values. In these cases, it may be most valid to fill the missing values in the column with a value calculated from the entries present in the column.</p>

Instructions 1/3
<p>Print the first five rows of the <code>StackOverflowJobsRecommend</code> column of <code>so_survey_df</code>.</p>

In [36]:
# Print the first five rows of StackOverflowJobsRecommend column
print(so_survey_df['StackOverflowJobsRecommend'].head())

0    NaN
1    7.0
2    8.0
3    NaN
4    8.0
Name: StackOverflowJobsRecommend, dtype: float64


Instructions 2/3
<p>Replace the missing values in the <code>StackOverflowJobsRecommend</code> column with its mean. Make changes directly to the original DataFrame.</p>

In [37]:
# Fill missing values with the mean
so_survey_df['StackOverflowJobsRecommend'].fillna(so_survey_df['StackOverflowJobsRecommend'].mean(), inplace=True)

# Print the first five rows of StackOverflowJobsRecommend column
print(so_survey_df['StackOverflowJobsRecommend'].head())

0    7.061602
1    7.000000
2    8.000000
3    7.061602
4    8.000000
Name: StackOverflowJobsRecommend, dtype: float64


Instructions 3/3
<p>Round the decimal values that you introduced in the <code>StackOverflowJobsRecommend</code> column.</p>

In [38]:
# Round the StackOverflowJobsRecommend values
so_survey_df['StackOverflowJobsRecommend'] = round(so_survey_df['StackOverflowJobsRecommend'])

# Print the top 5 rows
print(so_survey_df['StackOverflowJobsRecommend'].head())

0    7.0
1    7.0
2    8.0
3    7.0
4    8.0
Name: StackOverflowJobsRecommend, dtype: float64


**remember you should only round your values if you are certain it is applicable.**

### Imputing values in predictive models

<p>When working with predictive models you will often have a separate train and test DataFrames. In these cases you want to ensure no information from your test set leaks into your train set. When filling missing values in data to be used in these situations how should approach the two data sets?</p>

<pre>
Possible Answers

Only fill the train set.

Only fill the test set.

<b>Apply the measures of central tendency (mean/median etc.) calculated on the train set to both the train and test sets.</b>

Apply the measures of central tendency (mean/median etc.) calculated on the test set to both the train and test sets.

Apply the measures of central tendency (mean/median etc.) calculated on the train set to the train set, and the measures calculated on the test set, to the test set.

</pre>

**values calculated on the train test should be applied to both DataFrames.**

## Dealing with other data issues

### Dealing with stray characters (I)

<p>In this exercise, you will work with the <code>RawSalary</code> column of <code>so_survey_df</code> which contains the wages of the respondents along with the currency symbols and commas, such as <em>$42,000</em>. When importing data from Microsoft Excel, more often that not you will come across data in this form.</p>

Instructions 1/2
<p>Remove the commas (<code>,</code>) from the <code>RawSalary</code> column.</p>

In [39]:
# Remove the commas in the column
so_survey_df['RawSalary'] = so_survey_df['RawSalary'].str.replace(',', '')

Instructions 2/2
<p>Remove the dollar (<code>$</code>) signs from the <code>RawSalary</code> column.</p>

In [40]:
# Remove the dollar signs in the column
so_survey_df['RawSalary'] = so_survey_df['RawSalary'].str.replace('$', '')

**Replacing/removing specific characters is a very useful skill.**

### Dealing with stray characters (II)

<div class=""><p>In the last exercise, you could tell quickly based off of the <code>df.head()</code> call which characters were causing an issue. In many cases this will not be so apparent. There will often be values deep within a column that are preventing you from casting a column as a numeric type so that it can be used in a model or further feature engineering.</p>
<p>One approach to finding these values is to force the column to the data type desired using <code>pd.to_numeric()</code>, coercing any values causing issues to NaN, Then filtering the DataFrame by just the rows containing the NaN values.</p>
<p>Try to cast the <code>RawSalary</code> column as a float and it will fail as an additional character can now be found in it. Find the character and remove it so the column can be cast as a float.</p></div>

Instructions 1/2
<ul>
<li>Attempt to convert the <code>RawSalary</code> column of <code>so_survey_df</code> to numeric values coercing all failures into null values. </li>
<li>Find the indexes of the rows containing <code>NaN</code>s. </li>
<li>Print the rows in <code>RawSalary</code> based on these indexes.</li>
</ul>

In [41]:
# Attempt to convert the column to numeric values
numeric_vals = pd.to_numeric(so_survey_df['RawSalary'], errors='coerce')

# Find the indexes of missing values
idx = numeric_vals.isna()

# Print the relevant rows
print(so_survey_df['RawSalary'][idx])

0            NaN
2            NaN
4      £41671.00
6            NaN
8            NaN
         ...    
989          NaN
990          NaN
992          NaN
994          NaN
997          NaN
Name: RawSalary, Length: 401, dtype: object


Instructions 2/2

In [42]:
# Replace the offending characters
so_survey_df['RawSalary'] = so_survey_df['RawSalary'].str.replace('£', '')

# Convert the column to float
so_survey_df['RawSalary'] = so_survey_df['RawSalary'].astype(float)

# Print the column
print(so_survey_df['RawSalary'])

0            NaN
1        70841.0
2            NaN
3        21426.0
4        41671.0
         ...    
994          NaN
995      58746.0
996      55000.0
997          NaN
998    1000000.0
Name: RawSalary, Length: 999, dtype: float64


**Remember that even after removing all the relevant characters, you still need to change the type of the column to numeric if you want to plot these continuous values.**

### Method chaining

<div class=""><p>When applying multiple operations on the same column (like in the previous exercises), you made the changes in several steps, assigning the results back in each step. However, when applying multiple successive operations on the same column, you can "chain" these operations together for clarity and ease of management. This can be achieved by calling multiple methods sequentially:  </p>
<pre><code># Method chaining
df['column'] = df['column'].method1().method2().method3()

# Same as 
df['column'] = df['column'].method1()
df['column'] = df['column'].method2()
df['column'] = df['column'].method3()
</code></pre>
<p>In this exercise you will repeat the steps you performed in the last two exercises, but do so using method chaining.</p></div>

In [45]:
so_survey_df = pd.read_csv('https://raw.githubusercontent.com/lnunesAI/Datacamp/main/2-machine-learning-scientist-with-python/10-feature-engineering-for-machine-learning-in-python/datasets/so_survey_df.csv')

Instructions
<ul>
<li>Remove the commas (<code>,</code>) from the <code>RawSalary</code> column of <code>so_survey_df</code>. </li>
<li>Remove the dollar (<code>$</code>) signs from the <code>RawSalary</code> column. </li>
<li>Remove the pound (<code>£</code>) signs from the <code>RawSalary</code> column. </li>
<li>Convert the <code>RawSalary</code> column to float.</li>
</ul>

In [46]:
# Use method chaining
so_survey_df['RawSalary'] = so_survey_df['RawSalary']\
                              .str.replace(',', '')\
                              .str.replace('$', '')\
                              .str.replace('£', '')\
                              .astype(float)
 
# Print the RawSalary column
print(so_survey_df['RawSalary'])

0            NaN
1        70841.0
2            NaN
3        21426.0
4        41671.0
         ...    
994          NaN
995      58746.0
996      55000.0
997          NaN
998    1000000.0
Name: RawSalary, Length: 999, dtype: float64


**Custom functions can be also used when method chaining using the .apply() method.**